tensorflow_addons/losses/giou_loss.py

```diff
@@ -0,0 +1,141 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Implements GIoU loss."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import tensorflow as tf
+
+
+@tf.keras.utils.register_keras_serializable(package='Addons')
+class GIoULoss(tf.keras.losses.Loss):
+    """Implements the GIoU loss function.
+
+    GIoU loss was first introduced in the
+    [Generalized Intersection over Union paper](https://giou.stanford.edu/GIoU.pdf).
+    GIoU is a enhance for model which use IOU in object detection.
+
+    Usage:
+
+    ```python
+    gl = tfa.losses.GIoU()
+    boxes1 = tf.constant([[4.0, 3.0, 7.0, 5.0], [5.0, 6.0, 10.0, 7.0]])
+    boxes2 = tf.constant([[3.0, 4.0, 6.0, 8.0], [14.0, 14.0, 15.0, 15.0]])
+    loss = gl(boxes1,boxes2)
+    print('Loss: ', loss.numpy())  # Loss: [1.07500000298023224, 1.9333333373069763]
+    ```
+    Usage with tf.keras API:
+
+    ```python
+    model = tf.keras.Model(inputs, outputs)
+    model.compile('sgd', loss=tf.keras.losses.GIoULoss())
+    ```
+
+    Args:
+      mode: one of ['giou', 'iou'], decided to calculate giou loss or iou loss.
+    """
+
+    def __init__(self,
+                 mode='giou',
+                 reduction=tf.keras.losses.Reduction.AUTO,
+                 name='giou_loss'):
+        if mode not in ['giou', 'iou']:
+            raise ValueError("Value of mode should be 'iou' or 'giou'")
+        super(GIoULoss, self).__init__(name=name, reduction=reduction)
+        self.mode = mode
+
+    def get_config(self):
+        base_config = super(GIoULoss, self).get_config()
+        base_config['mode'] = self.mode
+        return base_config
+
+    def call(self, y_true, y_pred):
+        return giou_loss(y_true, y_pred, mode=self.mode)
+
+
+@tf.keras.utils.register_keras_serializable(package='Addons')
+@tf.function
+def giou_loss(y_true, y_pred, mode='giou'):
+    """
+    Args:
+        y_true: true targets tensor. The coordinates of the each bounding
+        box in boxes are encoded as [y_min, x_min, y_max, x_max].
+        y_pred: predictions tensor. The coordinates of the each bounding
+        box in boxes are encoded as [y_min, x_min, y_max, x_max].
+        mode: one of ['giou', 'iou'],
+        decided to calculate giou loss or iou loss.
+
+    Returns:
+        GIoU loss float `Tensor`.
+    """
+    # if y_true or y_pred:
+    if mode not in ['giou', 'iou']:
+        raise ValueError("Value of mode should be 'iou' or 'giou'")
+    y_pred = tf.convert_to_tensor(y_pred)
+    y_true = tf.cast(y_true, y_pred.dtype)
+    giou = do_giou_calculate(y_pred, y_true, mode)
+
+    # compute the final loss and return
+    return 1 - giou
+
+
+def do_giou_calculate(b1, b2, mode='giou'):
```

Comment:
Could we make this function private? Like `_calculate_giou`.

Reply:
It's a common function, maybe I should migrate it to tfa.image module?

Reply:
Umm, I would suggest leave it private here; let's move it in the future if there is any request!